Implement personal_sign [1/2]: generalize signing requests.

mandrigin opened the pull request:
We need to be able to sign more than just transactions to make DApps
work properly. This change separates signing requests from the
transactions and make it more general to prepare to intoduce different
types of signing requests.

This change is designed to preserve status APIs, so it is
backward-comparible with the current API bindings.

`personal_sign` implementation will follow in a separate PR.

#### TODO
- [x] CI: Fix compilation issues in `lib/library_test_utils.go`

adriacidre commented on geth/api/api.go:
```diff
@@ -49,9 +50,9 @@ func (api *StatusAPI) JailManager() jail.Manager {
-// TxQueueManager returns reference to account manager
-func (api *StatusAPI) TxQueueManager() *transactions.Manager {
-	return api.b.TxQueueManager()
+// TxFactory returns reference to account manager
```
not yours, but the comment says `returns reference to account manager` should say transaction factory instead?

adriacidre commented on geth/api/backend.go:
```diff
@@ -273,13 +264,28 @@ func (b *StatusBackend) registerHandlers() error {
 	}
-	rpcClient.RegisterHandler("eth_accounts", func(context.Context, ...interface{}) (interface{}, error) {
+	rpcClient.RegisterHandler(params.PersonalSignMethodName, func(context.Context, ...interface{}) (interface{}, error) {
+		return nil, errors.New("IGORM: custom erro boi from an override!")
```
What does that mean`IGORM: custom erro boi from an override!` ?

mandrigin replied:
oops, that is a leftover, this method shouldn't be even present in this PR because it breaks the APIs

adriacidre commented on geth/transactions/factory.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+// SetNetwork selects a correct network and RPC client.
+func (f *Factory) SetNetwork(networkID uint64, rpcClient *rpc.Client) {
```
Shouldn't be better to have 2 different setters instead, or have a name like `SetNetworkAndRPC`?

adriacidre commented on geth/transactions/factory.go:
```diff
@@ -0,0 +1,188 @@
+
+// SetNetwork selects a correct network and RPC client.
+func (f *Factory) SetNetwork(networkID uint64, rpcClient *rpc.Client) {
+	f.log.Info("start Manager")
```
Maybe this log message does not make sense anymore

adriacidre commented on geth/transactions/factory.go:
```diff
@@ -0,0 +1,188 @@
+		return account.ErrNoAccountSelected
+	}
+
+	// make sure that only account which created the tx can complete it
```
you probably don't need this comment anymore, it's already been set at the method's header.

divan commented on geth/api/backend.go:
```diff
@@ -273,13 +264,28 @@ func (b *StatusBackend) registerHandlers() error {
 	}
-	rpcClient.RegisterHandler("eth_accounts", func(context.Context, ...interface{}) (interface{}, error) {
+	rpcClient.RegisterHandler(params.PersonalSignMethodName, func(context.Context, ...interface{}) (interface{}, error) {
+		return nil, errors.New("IGORM: custom erro boi from an override!")
```
IGORM in an error description? :) Also, `erroboi` is a typo?
Maybe just "TBD" or "TODO"?

mandrigin replied:
Lol, missed that one! Thanks!

adriacidre commented on sign/README.md:
```diff
@@ -0,0 +1,13 @@
+Client has a chance to approve the sign request (by providing a valid password)
+or to discard it. When the request is approved, the locked functinality is
+executed.
+
```
Shouldn't README packages should be  generated with autoreadme like on [geth/jail package](https://github.com/status-im/status-go/blob/develop/geth/jail/doc.go)

mandrigin replied:
good point

mandrigin replied:
@adriacidre hmm, it looks like we don't have a makefile target or something to do that.
let's skip it for this PR then.

divan commented on geth/transactions/factory.go:
```diff
@@ -0,0 +1,188 @@
+)
+
+// Factory validates, signs and propagates transactions to the Ethereum network.
+type Factory struct {
```
I'm afraid the word "Factory" has a very strong association with Factory pattern from class-based programming and does not really represent the purpose of this type.
As much as `-Manager` is bad, it's still better describes what this abstraction is for.

mandrigin replied:
I don't know. This class doesn't manage anything, it creates and propagates the transactions to the Ethereum network... But I think you're right, Factory is a worse of a name.

themue replied:
`Transactor` 🤖

mandrigin replied:
Yeah, that might make sense. `Transactor#SendTransaction` and `Transaction#validateAndPropagate` make sense to me.

divan commented on geth/transactions/factory.go:
```diff
@@ -0,0 +1,188 @@
+	return &Factory{
+		signRequests:      signRequests,
+		addrLock:          &AddrLocker{},
+		completionTimeout: sendTransactionTimeout,
```
Maybe use the same name here, either `completionTImeout` or `sendTransactionTimeout`?

divan commented on geth/transactions/factory.go:
```diff
@@ -0,0 +1,188 @@
+		completionTimeout: sendTransactionTimeout,
+		rpcCallTimeout:    rpcCallTimeout,
+		localNonce:        sync.Map{},
+		log:               log.New("package", "status-go/geth/transactions.Manager"),
```
Are we using this approach (injecting logger) throughout the code? I'm not against this approach but it's better to be consistent and use the same approach in all packages (i.e. using global logger vs injecting logger).

mandrigin replied:
yeap, we do that since we removed our log package in favour of ethereum log. so we decided to follow their patterns too.

divan commented on geth/transactions/factory.go:
```diff
@@ -0,0 +1,188 @@
+	// make sure that only account which created the tx can complete it
+	if args.From.Hex() != selectedAccount.Address.Hex() {
+		err := sign.ErrInvalidCompleteTxSender
+		f.log.Warn("queued transaction does not belong to the selected account", "err", err)
```
Let's try to avoid using warning level. It's an error.

mandrigin replied:
ah, that's a leftover from the initial implementation. that'd be visible after the file is renamed back :)

adriacidre commented on sign/pending_requests.go:
```diff
@@ -0,0 +1,186 @@
+func NewRequests() *PendingRequests {
+	logger := log.New("package", "status-go/sign.PendingRequests")
+
+	logger.Info("initializing transaction queue")
```
is this log still relevant?

mandrigin replied:
nope

adriacidre commented on sign/pending_requests.go:
```diff
@@ -0,0 +1,186 @@
+	return nil
+}
+
+// Approves a signing request by it's ID. Requires a valid password and a verification function.
```
Lint : `Approve` instead of `Approves`

dshulyak commented:
@mandrigin can you please do a file copy with `git mv`? it is somewhat hard to see actual changes this way

divan commented on geth/transactions/types.go:
```diff
@@ -68,3 +51,31 @@ func (args SendTxArgs) GetInput() hexutil.Bytes {
+	rpcCall := rpc.Call{Params: args}
+	fromAddr, err = rpcCall.ParseFromAddress()
+	if err != nil {
+		fromAddr = common.HexToAddress("0x0")
```
Shouldn't we abort the transaction in the case of wrong/incorrect from/to addresses?

mandrigin replied:
that code is actually untouched by me. I'll take a look if I can make diffs smaller, because this shouldn't be removed/added.

dshulyak commented on sign/pending_requests.go:
```diff
@@ -0,0 +1,186 @@
+	rs.requests[request.ID] = request
+	rs.log.Info("signing request is created", "ID", request.ID)
+
+	go NotifyOnEnqueue(request.ID, meta, ctx)
```
i think we are trying to move notifications as close to backend as possible

mandrigin replied:
well, it is as close to the backend as it was before.
I want to extract these signals to a separate entity and use it as a dependency, but that PR grew too big anyway.

dshulyak replied:
it was in transactor code, which is a bit closer i guess.
still, why to have it here and not in the caller?

why do we need to spawn goroutine?

mandrigin replied:
Essentially, you are subscribing to the requests collection events, if a sign request was created or errored, that what the API consumer (aka status-react) should know for now. It's nothing to do specifically for transactions, it just so happens that it is used to be that way. But we are really interested in double-checking the user's password on some occasions. It might be signing a transaction, it might be not :)

And that is as close (PendingRequests) is a member of a Backend anyway.

Spawning a goroutine here is a cheap an simple way to guarantee that any notification handler won't deadlock this code by, for instance, waiting for the request to be fulfilled inside.

adriacidre reviewed (commented):
Awesome job!
A part of minor comments, why does this package lives outside `geth`?

divan reviewed (commented):
Terrific job! See my comments below.

mandrigin commented:
@dshulyak I just tried `git mv` but for files that are heavily modified it still shows `deleted/added`.

mandrigin commented:
@adriacidre 
> why does this package lives outside geth?

We wanted to get rid of `geth` package, it doesn't really bring any meaning. So, I decided to keep new features outside of it.

adambabik commented on geth/transactions/manager.go:
```diff
@@ -0,0 +1,186 @@
+
+// Manager validates, signs and propagates transactions to the Ethereum network.
+type Manager struct {
+	signRequests   *sign.PendingRequests
```
Can we unify this? Any particular reason why the filed is called `signRequests` instead of `pendingRequests`?

adambabik commented on sign/pending_requests.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+// Reset is to be used in tests only, as it simply creates new transaction map, w/o any cleanup of the previous one
+func (rs *PendingRequests) Reset() {
```
Why `rs` instead of `r`?

adambabik replied:
Not sure why we need to expose it if we can create a new `PendingRequests` object in tests.

mandrigin replied:
Agree about exposing.
`rs` is a typical (for me) shortcut for a collection.
like `t` is `Transaction` `ts` is `Transactions`.

mandrigin commented:
ok, I think I fixed the review comments :)

adambabik commented on geth/api/backend.go:
```diff
@@ -34,8 +35,9 @@ var (
 type StatusBackend struct {
 	mu              sync.Mutex
 	statusNode      *node.StatusNode
+	signRequests    *sign.PendingRequests
```
It would be better to use a single name. Currently, there are `signRequests` and `sign.PendingRequests` which look like equivalents. Would it be possible to unify them, for instance: `signRequests: *sign.Requests` or `pendingRequests: *sign.PendingRequests`?

mandrigin replied:
Yep, I'll take this one: `pendingRequests` / `*sign.PendingRequests`

adambabik commented on geth/transactions/geth_transactor.go:
```diff
@@ -12,37 +12,28 @@ import (
-	ethereum.GasPricer
-	ethereum.TransactionSender
+// gethTransactor wraps upstream RPC APIs related to transactions
+type gethTransactor struct {
```
Any particular reason to prefix it with `geth`? Also, it's just a wrapper for an RPC client. Maybe `transactionClient`?

adambabik commented on geth/transactions/geth_transactor.go:
```diff
@@ -12,37 +12,28 @@ import (
 // PendingNonceAt returns the account nonce of the given account in the pending state.
 // This is the nonce that should be used for the next transaction.
-func (ec *EthTxClient) PendingNonceAt(ctx context.Context, account common.Address) (uint64, error) {
+func (gt *gethTransactor) PendingNonceAt(ctx context.Context, account common.Address) (uint64, error) {
```
if we refactor that, maybe we can stay with a single char: `t *gethTransactor`?

mandrigin replied:
`rpcTransactor` i guess then...

adambabik commented on geth/transactions/transactor.go:
```diff
@@ -0,0 +1,187 @@
+// It uses upstream to propagate transactions to the Ethereum network.
+type Transactor struct {
+	signRequests   *sign.PendingRequests
+	upstream       *gethTransactor
```
It's not always upstream. If upstream node is not used, it's in-proc RPC client.

mandrigin replied:
It's upstream to this particular struct, not to the whole node. In-proc RPC is talking to something external, otherwise, we wouldn't use RPC, right? :)
It doesn't really matter if the node happens to run in the same process.
But sure, I'm up for the name suggestions that gives more self-explanatory lines of code than
`t.upstream.SendTransaction` down in this class :)

mandrigin replied:
thanks!

adambabik commented on geth/transactions/transactor.go:
```diff
@@ -0,0 +1,187 @@
+// Transactor validates, signs transactions.
+// It uses upstream to propagate transactions to the Ethereum network.
+type Transactor struct {
+	signRequests   *sign.PendingRequests
```
`signRequests` vs `PendingRequests` confusion.

adambabik commented on geth/transactions/transactor.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+// SetUpstream selects a correct network and RPC client.
+func (t *Transactor) SetUpstream(networkID uint64, rpcClient *rpc.Client) {
```
`SetUpstream` is not accurate as `rpcClient` can be only in-proc RPC client. Recommend splitting it into `SetNetworkID` and `SetRPCClient`.

mandrigin replied:
For this transactor it is upstream, we forward `eth_sendTransaction` somewhere, that could be an upstream node or our local node. Both are external to this one.

adambabik commented on sign/pending_requests.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+// First returns a first signing request (if exists, nil otherwise). Used in tests.
+func (rs *PendingRequests) First() *Request {
```
Again here, I would avoid creating methods just for the purposes of the test. This comment may quickly become invalid when someone uses this method in a different place.

mandrigin replied:
I'd keep `r` for `Request` and `rs` for lists/dictionaries, that are `Request*s*`

themue replied:
That's my convention too.

mandrigin replied:
@adambabik I will fix the comment. we have quite a few places where we basically have test-only methods in this class. I wanted to avoid refactoring tests too much... Maybe when we extract notifications into something, it would be much easier.

adriacidre commented:
Looks good so far, but, there are some linter errors on [travis](https://travis-ci.org/status-im/status-go/builds/363089108?utm_source=github_status&utm_medium=notification)

adambabik commented on sign/request.go:
```diff
@@ -0,0 +1,36 @@
+		context:      ctx,
+		locked:       false,
+		completeFunc: completeFunc,
+		result:       make(chan Result, 1),
```
If a buffered channel is used, then `go func() { request.result <- result }()` does not need to be called in a goroutine. Personally, I prefer unbuffered channels so I would change this line.

dshulyak replied:
why do you prefer them? they actually often lead to unnecessary contention

adambabik replied:
Because if the business logic is wrong, the same problem will be experienced with buffered channels just a bit later. Buffered channels are cool if there is a typical producer-consumer problem but I don't see it to be the case here.

adambabik commented on sign/result.go:
```diff
@@ -0,0 +1,9 @@
+
+import "github.com/ethereum/go-ethereum/common"
+
+// Result is a result of a signing request, error or successful:w
```
`:w` leftover from vim? 😄

adambabik reviewed (commented):
I added a few comments but it's a very good direction and great work 👏

dshulyak commented on sign/pending_requests_test.go:
```diff
@@ -0,0 +1,203 @@
+
+type PendingRequestsSuite struct {
+	suite.Suite
+	queue *PendingRequests
```
not a queue

dshulyak commented on static/testdata/dapps/personal_sign.js:
```diff
@@ -0,0 +1,22 @@
+console.log("Hellow from static!");
```
is this file added intentionally?

mandrigin replied:
nope, thanks!

dshulyak commented on sign/pending_requests.go:
```diff
@@ -0,0 +1,185 @@
+	// sending result to the request in a separate goroutine to avoid deadlocks if
+	// both `Wait` and `Complete` are called from the same goroutine
+	// (they are in the tests, for instance).
+	go func() { request.result <- result }()
```
it wasn't in a goroutine in previous code, right?

mandrigin replied:
Well, that is complicated, because the previous code did essentially everything through channelling. But I might try to remove that and see if your tests won't deadlock :)

dshulyak commented on sign/pending_requests.go:
```diff
@@ -0,0 +1,185 @@
+
+	request.locked = false
+
+	go NotifyOnReturn(request.ID, request.Meta, request.context, err)
```
why do we need to spawn goroutine?

dshulyak commented on geth/transactions/geth_transactor.go:
```diff
@@ -12,37 +12,28 @@ import (
 )
 
-// EthTransactor provides methods to create transactions for ethereum network.
-type EthTransactor interface {
```
why interface disappeared? you could for example just set simulated backend as eth tx client in tx queue manager and do an advanced tests, or replace our geth wrapper with regular geth rpc client without tweaking codebase

mandrigin replied:
well, when we do that, we will add the interface, it's pretty obvious. before that it is a dead code that serves no purpose.
this dependency is created inside the `PendingRequests`, it isn't injected from outside. so fiddling with the code would be necessary.

dshulyak replied:
can you please keep it? i started to use it in tests for this PR https://github.com/status-im/status-go/pull/807. and i will have to add it back then

dshulyak commented on sign/pending_requests_test.go:
```diff
@@ -0,0 +1,203 @@
+
+var defaultCompleteError = errors.New("default completion error")
+
+const correctPassword = "password-correct"
```
const (
...
)

dshulyak reviewed (changes requested):
mainly nitpicks, can be ignored. though there are couple of things:
- i am not sure if that's good to spawn goroutines for a task that will take millisecond, they are cheap but not free
- changes to bindata seem to be out of place

mandrigin commented:
@dshulyak thought about goroutines here in general. re-asking for a user of his password is a relatively rare task. Per each request, we might have min 2 goroutines (enqueue notification + result passing) or 2 + n notifications for a wrong password entry. So all these goroutines are most likely to be very sparsely distributed in time.
Having them protects from deadlocks, which is important since we block a thread in some of our methods. That is a very simple implementation of such protection and not needing to write that we shouldn't ever call `Wait` in a wrong goroutine.

mandrigin committed:
Generalize signing requests.

We need to be able to sign more than just transactions to make DApps
work properly. This change separates signing requests from the
transactions and make it more general to prepare to intoduce different
types of signing requests.

This change is designed to preserve status APIs, so it is
backward-comparible with the current API bindings.

dshulyak commented:
> Having them protects from deadlocks, which is important since we block a thread in some of our methods. That is a very simple implementation of such protection and not needing to write that we shouldn't ever call Wait in a wrong goroutine.

is it possible with buffered chan to get such deadlock?

also, there are 2 other goroutines for notifications

mandrigin commented:
@dshulyak I was actually talking about notifications 😅 they provide 1+n goroutines per a sign request if we are unlucky. For the result, I'll try to remove this goroutine, it should be fine.

adambabik commented on geth/transactions/transactor.go:
```diff
@@ -0,0 +1,200 @@
+// It uses upstream to propagate transactions to the Ethereum network.
+type Transactor struct {
+	pendingSignRequests *sign.PendingRequests
+	upstream            ethTransactor
```
Is it a leftover because `upstream` does not make sense here. This is getting a bit tricky because we have `Transactor` here and `ethTransactor` in a different place which both do not really relate to each other...

mandrigin replied:
It's all very confusing in the class responsibilities too.
We use our own Transactor to override *some* behaviour of how original transactions mechanics work in go-ethereum.
We override 2 things there: 
1) we provide a halt-and-wait notifications for the API clients to be able to confirm or decline the transaction
2) we provide a local nonce algo that @dshulyak made

so there we have it, we have all the Transactors implementing `SendTransaction` method, so we can make an interface out of it, for sure.

but to function properly our `transactions.Transactor` needs a bit more API than just `SendTransaction`. (we need it for local nonce mostly, afaik).
maybe we need to fiddle on that naming too.

mandrigin replied:
hmm, I guess I found a better scheme

mandrigin commented:
Rebased to the latest `develop` and addressed the review issues. I got another idea about how to make Transactor/RPC entities more obvious, but I don't really want to keep it to the same PR.
The same for notifications refactoring.

adambabik commented on geth/transactions/rpc_transactor.go:
```diff
@@ -9,40 +9,40 @@ import (
-	c *rpc.Client
+// rpcTransactor wraps upstream RPC APIs related to transactions
+type rpcTransactor struct {
+	rpc *rpc.Client
```
It's ok but I would go for `client` as `rpc` is much broader in meaning.

mandrigin replied:
it should be `rpcClient`, then 😄

mandrigin commented:
@adambabik @dshulyak hopefully, I addressed everything now :)

dshulyak commented on geth/transactions/transactor.go:
```diff
@@ -0,0 +1,205 @@
+// SetRPCClient an RPC client.
+func (t *Transactor) SetRPCClient(rpcClient *rpc.Client) {
+	rpcWrapper := newRPCWrapper(rpcClient)
+	t.sender = rpcWrapper
```
what is the reason to split?

mandrigin replied:
to avoid many responsibilities through the same interface. there are clearly 3 groups of them:
- the one that sends transactions (potentially we can even make our `Transactor` comply to the same interface;
- the one that provides information about gas prices;
- the one that provides information about pending nonce.

When I grouped all three of these, I really struggle with naming them anything except, well, a manager, a wrapper or a helper. It isn't really a transactor, because our transactor doesn't comply.

The benefits I see:
- simple naming and clear responsibilities;
- the code inside `verifyAndPropagate` now has constructions like `sender.SendTransaction` and `gasCalculator.EstimateGasPrice` which are very readable IMO;
- more flexibility for potential testing if we want to test only a part of this interface.

The downside:
- more interfaces
- more fields

I think it worth the deal.

dshulyak replied:
not sure, usually, it is the same object that provides all of this methods

mandrigin replied:
maybe it is too java-ish, but still :)
I'm struggling with naming everything consistently here, so we don't end up with 2 different transactors like @adambabik pointed out.

mandrigin replied:
Since looks like this is the only part left.
I'd maybe leave it for another thought later.
These names are still better than `TxQueueManager`, IMO

dshulyak reviewed (approved):
i can't find what was the reason to split eth transactor interface?

JekaMas reviewed (approved):
impressive.
